test/Test/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -0,0 +1,338 @@
+  observeModel = \case
+    MEntry e -> OId e
+    MKey k -> OId k
+    MHasMore h -> OId h
+    MInt n -> OId n
+    MUnit () -> OId ()
+    MEither x -> OEither $ bimap observeModel observeModel x
+    MTuple2 x -> OTuple2 $ bimap observeModel observeModel x
+    MTuple3 x -> OTuple3 $ trimap observeModel observeModel observeModel x
+
+  modelNextState :: forall a.
+       LockstepAction ReadersState a
+    -> ModelLookUp ReadersState
+    -> ReadersState
+    -> (ReadersVal a, ReadersState)
+  modelNextState action lookUp (ReadersState mock) =
+      ReadersState <$> runMock lookUp action mock
+
+  usedVars :: LockstepAction ReadersState a -> [AnyGVar (ModelOp ReadersState)]
+  usedVars = const []
+
+  arbitraryWithVars ::
+       ModelFindVariables ReadersState
+    -> ReadersState
+    -> Gen (Any (LockstepAction ReadersState))
+  arbitraryWithVars _ (ReadersState mock)
+    | isEmpty mock =
+        -- It's not allowed to keep using a drained RunReaders,
+        -- we can only create a new one.
+        Some . New <$> (vector =<< chooseInt (1, 10))
+    | otherwise =
+        QC.frequency $
+          [ (3, pure (Some PeekKey))
+          , (5, pure (Some (Pop 1)))
+          , (1, Some . Pop <$> chooseInt (1, size mock))  -- drain a significant amount
+          , (1, Some . DropWhileKey <$> arbitrary)  -- most likely nothing to drop
+          ] <>
+          [ (3, pure (Some (DropWhileKey k)))  -- drops at least one key
+          | Right k <- [peekKeyMock mock]
+          ]
+
+  shrinkWithVars ::
+       ModelFindVariables ReadersState
+    -> ReadersState
+    -> LockstepAction ReadersState a
+    -> [Any (LockstepAction ReadersState)]
+  shrinkWithVars _ _ = \case
+      New wbs -> Some . New <$> shrink wbs
+      Pop n   -> Some . Pop <$> shrink n
+      _       -> []
+
+  -- TODO: tag, e.g.
+  -- - fully draining the Readers
+  -- - DropWhileKey dropping multiple entries
+  tagStep _ _ _ = []
+
+runMock ::
+     lookUp
+  -> Action (Lockstep ReadersState) a
+  -> MockReaders
+  -> (ReadersVal a, MockReaders)
+runMock _ = \case
+    New wbs        -> const $ wrap MUnit (Right (), newMock (map unTypedWriteBuffer wbs))
+    PeekKey        -> \m -> wrap MKey (peekKeyMock m, m)
+    Pop n          -> wrap wrapPop . popMock n
+    DropWhileKey k -> wrap wrapDrop . dropWhileKeyMock k
+  where
+    wrap :: (a -> ReadersVal b) -> (Either () a, MockReaders) -> (ReadersVal (Either () b), MockReaders)
+    wrap f = first (MEither . bimap MUnit f)
+
+    wrapPop = MTuple3 . trimap MKey MEntry MHasMore
+
+    wrapDrop = MTuple2 . bimap MInt MHasMore
+
+trimap :: (a -> a') -> (b -> b') -> (c -> c') -> (a, b, c) -> (a', b', c')
+trimap f g h (a, b, c) = (f a, g b, h c)
+
+data Op a b where
+  OpId   :: Op a a
+  OpUnit :: Op a ()
+
+instance Operation Op where
+  opIdentity = OpId
+
+instance InterpretOp Op (ModelValue ReadersState) where
+  intOp OpId   x = Just x
+  intOp OpUnit _ = Just (MUnit ())
```

Comment:
My intuition says that this isn't needed, and you can use the existing [SumProd](https://hackage.haskell.org/package/quickcheck-lockstep-0.5.0/docs/Test-QuickCheck-StateModel-Lockstep-Op-SumProd.html) instead

Reply:
In fact I wasn't using the type at all, just deleted it. :smile: